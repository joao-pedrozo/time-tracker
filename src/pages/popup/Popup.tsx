import { useEffect, useState } from "react";
import WebsitesList from "./WebsitesList";
import WebsitesChart from "./WebsitesChart";
import Header from "./Header";

export interface SitesData {
  url: string;
  lastRecorded: Date;
  days: {
    date: string;
    timeSpent: number;
  }[];
}

const blacklistedSites = ["extensions", "newtab"];

export default function Popup(): JSX.Element {
  const [sitesData, setSitesData] = useState<SitesData[]>([]);
  const [currentDay, setCurrentDay] = useState<string>("");

  useEffect(() => {
    chrome.storage.local.get("sitesData", (data) => {
      console.log(data);

      if (data && data.sitesData) {
        setSitesData(data.sitesData);
        setCurrentDay(data.sitesData[0]?.days[0]?.date || "");
      }
    });
  }, []);

  const handlePreviousDay = () => {
    const currentIndex = sitesData.findIndex((site) =>
      site.days.some((day) => day.date === currentDay)
    );
    if (currentIndex > 0) {
      const previousDay = sitesData[currentIndex - 1].days[0].date;
      setCurrentDay(previousDay);
    }
  };

  const handleNextDay = () => {
    const currentIndex = sitesData.findIndex((site) =>
      site.days.some((day) => day.date === currentDay)
    );
    if (currentIndex < sitesData.length - 1) {
      const nextDay = sitesData[currentIndex + 1].days[0].date;
      setCurrentDay(nextDay);
    }
  };

  const filteredSitesData = sitesData
    .filter(
      (site) =>
        site.days.some((day) => day.date === currentDay) &&
        !blacklistedSites.some((blacklistedSite) =>
          site.url.includes(blacklistedSite)
        )
    )
    .map((site) => ({
      ...site,
      url: site.url.replace("www.", ""),
    }))
    .sort((a, b) => {
      const siteA = a.days.find((day) => day.date === currentDay);
      const siteB = b.days.find((day) => day.date === currentDay);

      if (!siteA || !siteB) {
        return 0;
      }

      return siteB.timeSpent - siteA.timeSpent;
    });

  return (
    <div className="flex items-center flex-col bg-white dark:dark:bg-[#0f0f0f]">
      <Header />
      <div className="flex justify-between items-center px-4 gap-2 text-white mt-[8px] mb-[18px]">
        <button onClick={handlePreviousDay}>Previous Day</button>
        <h2>{currentDay}</h2>
        <button onClick={handleNextDay}>Next Day</button>
      </div>
      {filteredSitesData.length > 0 && (
        <WebsitesChart sites={filteredSitesData} />
      )}
      {filteredSitesData.length > 0 && (
        <WebsitesList currentDay={currentDay} websites={filteredSitesData} />
      )}
    </div>
  );
}
