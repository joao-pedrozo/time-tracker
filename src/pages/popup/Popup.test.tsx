import { render, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import Popup from "./Popup";

const setStorage = vi.fn();
const getStorage = vi.fn();

(global.chrome as any) = {
  storage: {
    local: {
      get: getStorage,
      set: setStorage,
    },
  },
};

const mockWebsites = [
  {
    url: "https://www.google.com",
    lastRecorded: new Date(),
    days: [
      {
        date: "2022-01-01",
        timeSpent: 1000,
      },
    ],
  },
  {
    url: "https://www.facebook.com",
    lastRecorded: new Date(),
    days: [
      {
        date: "2022-01-01",
        timeSpent: 2000,
      },
    ],
  },
];

describe("Popup", () => {
  beforeEach(() => {
    getStorage.mockImplementation((_, cb) => cb({ sitesData: mockWebsites }));
  });

  it("should render correctly", async () => {
    const { findByText, container } = render(<Popup />);

    expect(await findByText("Webtime Tracker")).toBeInTheDocument();
  });

  it("should list all websites with their respective spent time", async () => {
    const { findByText, queryByText, findAllByRole } = render(<Popup />);

    expect(await findByText("https://google.com")).toBeInTheDocument();
    expect(await findByText("https://facebook.com")).toBeInTheDocument();
    expect(queryByText("https://twitter.com")).not.toBeInTheDocument();

    const listItems = await findAllByRole("listitem");

    const facebookElement = listItems.find((li) =>
      li.textContent?.includes("https://facebook.com")
    );
    const googleElement = listItems.find((li) =>
      li.textContent?.includes("https://google.com")
    );

    expect(facebookElement).toHaveTextContent("2s");
    expect(googleElement).toHaveTextContent("1s");
  });
});
