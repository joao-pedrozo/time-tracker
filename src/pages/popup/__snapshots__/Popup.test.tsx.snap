// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Popup > should render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex items-center flex-col bg-white dark:dark:bg-[#0f0f0f]"
      >
        <header
          class="w-full font-bold text-2xl flex justify-between items-center px-4 pt-4"
        >
          <h1
            class="bg-gradient-to-r from-green-400 to-blue-500 inline-block text-transparent bg-clip-text"
          >
            Webtime Tracker
          </h1>
          <div
            class="mt-[-8px] flex gap-2"
          >
            <button>
              <svg
                class="p-0 text-black dark:text-white"
                fill="currentColor"
                height="24"
                viewBox="0 0 256 256"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M240,96a8,8,0,0,1-8,8H216v16a8,8,0,0,1-16,0V104H184a8,8,0,0,1,0-16h16V72a8,8,0,0,1,16,0V88h16A8,8,0,0,1,240,96ZM144,56h8v8a8,8,0,0,0,16,0V56h8a8,8,0,0,0,0-16h-8V32a8,8,0,0,0-16,0v8h-8a8,8,0,0,0,0,16Zm72.77,97a8,8,0,0,1,1.43,8A96,96,0,1,1,95.07,37.8a8,8,0,0,1,10.6,9.06A88.07,88.07,0,0,0,209.14,150.33,8,8,0,0,1,216.77,153Zm-19.39,14.88c-1.79.09-3.59.14-5.38.14A104.11,104.11,0,0,1,88,64c0-1.79,0-3.59.14-5.38A80,80,0,1,0,197.38,167.86Z"
                />
              </svg>
            </button>
            <a
              href="https://github.com/joao-pedrozo/time-tracker"
              rel="noreferrer"
              target="_blank"
            >
              <svg
                class="p-0 text-black dark:text-white"
                fill="currentColor"
                height="24"
                viewBox="0 0 256 256"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M208.31,75.68A59.78,59.78,0,0,0,202.93,28,8,8,0,0,0,196,24a59.75,59.75,0,0,0-48,24H124A59.75,59.75,0,0,0,76,24a8,8,0,0,0-6.93,4,59.78,59.78,0,0,0-5.38,47.68A58.14,58.14,0,0,0,56,104v8a56.06,56.06,0,0,0,48.44,55.47A39.8,39.8,0,0,0,96,192v8H72a24,24,0,0,1-24-24A40,40,0,0,0,8,136a8,8,0,0,0,0,16,24,24,0,0,1,24,24,40,40,0,0,0,40,40H96v16a8,8,0,0,0,16,0V192a24,24,0,0,1,48,0v40a8,8,0,0,0,16,0V192a39.8,39.8,0,0,0-8.44-24.53A56.06,56.06,0,0,0,216,112v-8A58.14,58.14,0,0,0,208.31,75.68ZM200,112a40,40,0,0,1-40,40H112a40,40,0,0,1-40-40v-8a41.74,41.74,0,0,1,6.9-22.48A8,8,0,0,0,80,73.83a43.81,43.81,0,0,1,.79-33.58,43.88,43.88,0,0,1,32.32,20.06A8,8,0,0,0,119.82,64h32.35a8,8,0,0,0,6.74-3.69,43.87,43.87,0,0,1,32.32-20.06A43.81,43.81,0,0,1,192,73.83a8.09,8.09,0,0,0,1,7.65A41.72,41.72,0,0,1,200,104Z"
                />
              </svg>
            </a>
          </div>
        </header>
        <div
          class="flex justify-between items-center px-4 py-2 gap-2 text-white my-[9px]"
        >
          <button>
            Previous Day
          </button>
          <h2 />
          <button>
            Next Day
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex items-center flex-col bg-white dark:dark:bg-[#0f0f0f]"
    >
      <header
        class="w-full font-bold text-2xl flex justify-between items-center px-4 pt-4"
      >
        <h1
          class="bg-gradient-to-r from-green-400 to-blue-500 inline-block text-transparent bg-clip-text"
        >
          Webtime Tracker
        </h1>
        <div
          class="mt-[-8px] flex gap-2"
        >
          <button>
            <svg
              class="p-0 text-black dark:text-white"
              fill="currentColor"
              height="24"
              viewBox="0 0 256 256"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M240,96a8,8,0,0,1-8,8H216v16a8,8,0,0,1-16,0V104H184a8,8,0,0,1,0-16h16V72a8,8,0,0,1,16,0V88h16A8,8,0,0,1,240,96ZM144,56h8v8a8,8,0,0,0,16,0V56h8a8,8,0,0,0,0-16h-8V32a8,8,0,0,0-16,0v8h-8a8,8,0,0,0,0,16Zm72.77,97a8,8,0,0,1,1.43,8A96,96,0,1,1,95.07,37.8a8,8,0,0,1,10.6,9.06A88.07,88.07,0,0,0,209.14,150.33,8,8,0,0,1,216.77,153Zm-19.39,14.88c-1.79.09-3.59.14-5.38.14A104.11,104.11,0,0,1,88,64c0-1.79,0-3.59.14-5.38A80,80,0,1,0,197.38,167.86Z"
              />
            </svg>
          </button>
          <a
            href="https://github.com/joao-pedrozo/time-tracker"
            rel="noreferrer"
            target="_blank"
          >
            <svg
              class="p-0 text-black dark:text-white"
              fill="currentColor"
              height="24"
              viewBox="0 0 256 256"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M208.31,75.68A59.78,59.78,0,0,0,202.93,28,8,8,0,0,0,196,24a59.75,59.75,0,0,0-48,24H124A59.75,59.75,0,0,0,76,24a8,8,0,0,0-6.93,4,59.78,59.78,0,0,0-5.38,47.68A58.14,58.14,0,0,0,56,104v8a56.06,56.06,0,0,0,48.44,55.47A39.8,39.8,0,0,0,96,192v8H72a24,24,0,0,1-24-24A40,40,0,0,0,8,136a8,8,0,0,0,0,16,24,24,0,0,1,24,24,40,40,0,0,0,40,40H96v16a8,8,0,0,0,16,0V192a24,24,0,0,1,48,0v40a8,8,0,0,0,16,0V192a39.8,39.8,0,0,0-8.44-24.53A56.06,56.06,0,0,0,216,112v-8A58.14,58.14,0,0,0,208.31,75.68ZM200,112a40,40,0,0,1-40,40H112a40,40,0,0,1-40-40v-8a41.74,41.74,0,0,1,6.9-22.48A8,8,0,0,0,80,73.83a43.81,43.81,0,0,1,.79-33.58,43.88,43.88,0,0,1,32.32,20.06A8,8,0,0,0,119.82,64h32.35a8,8,0,0,0,6.74-3.69,43.87,43.87,0,0,1,32.32-20.06A43.81,43.81,0,0,1,192,73.83a8.09,8.09,0,0,0,1,7.65A41.72,41.72,0,0,1,200,104Z"
              />
            </svg>
          </a>
        </div>
      </header>
      <div
        class="flex justify-between items-center px-4 py-2 gap-2 text-white my-[9px]"
      >
        <button>
          Previous Day
        </button>
        <h2 />
        <button>
          Next Day
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
